File: internal/backup/backup.go
// Package backup implements opendray's two backup-shaped surfaces:
//
//   - Operator-facing disaster-recovery backups (A): scheduled or
//     manual full PostgreSQL dumps, encrypted with a passphrase
//     derived AES-256 key, written to a pluggable BackupTarget
//     (local disk, SMB, ...). State persists in the backups,
//     backup_schedules, and backup_targets tables (migration 0014).
//   - Admin-facing data exports (C): one-shot zip bundles of
//     selected logical entities (memories, integrations metadata,
//     custom tasks) downloaded once via a per-export token. State
//     persists in the exports table (migration 0015).
//
// Both surfaces share cipher / target / pgdump primitives but are
// otherwise independent — runtime flow lives in service_runtime.go,
// export flow in service_export.go.
package backup

import (
	"errors"
	"fmt"
	"time"
)

// BackupStatus is the lifecycle state of a single Backup row.
type BackupStatus string

const (
	BackupPending   BackupStatus = "pending"
	BackupRunning   BackupStatus = "running"
	BackupSucceeded BackupStatus = "succeeded"
	BackupFailed    BackupStatus = "failed"
	// BackupDeleted means the row is retained for audit but the
	// underlying blob has been removed from its target. Listing
	// endpoints filter these out by default.
	BackupDeleted BackupStatus = "deleted"
)

// TriggeredBy distinguishes how a backup run was started.
type TriggeredBy string

const (
	TriggeredScheduler TriggeredBy = "scheduler"
	TriggeredManual    TriggeredBy = "manual"
	TriggeredAPI       TriggeredBy = "api"
	// TriggeredPreMigrate marks a snapshot taken automatically just
	// before schema migrations run, so an upgrade is always preceded
	// by a restorable point.
	TriggeredPreMigrate TriggeredBy = "pre_migrate"
	// TriggeredPreRestore marks the safety snapshot taken automatically
	// before an apply-mode restore overwrites the current instance.
	TriggeredPreRestore TriggeredBy = "pre_restore"
)

// BackupKind is how much of an instance a backup captures.
type BackupKind string

const (
	// KindDBOnly is a plain encrypted pg_dump — the historical default.
	KindDBOnly BackupKind = "db_only"
	// KindFullInstance additionally bundles the vault (notes/skills/mcp),
	// secrets.env and config.toml: everything needed to rebuild a
	// working instance on a fresh machine, not just its database.
	KindFullInstance BackupKind = "full_instance"
)

// orDefault normalises an empty kind to the historical default so old
// rows and callers that don't set a kind keep behaving as db_only.
func (k BackupKind) orDefault() BackupKind {
	if k == "" {
		return KindDBOnly
	}
	return k
}

// ParseBackupKind validates a kind string from an API request,
// defaulting empty to db_only and rejecting unknown values.
func ParseBackupKind(s string) (BackupKind, error) {
	switch BackupKind(s) {
	case "", KindDBOnly:
		return KindDBOnly, nil
	case KindFullInstance:
		return KindFullInstance, nil
	default:
		return "", fmt.Errorf("backup: unknown kind %q", s)
	}
}

// TargetKind identifies the storage backend behind a BackupTarget.
type TargetKind string

const (
	TargetLocal  TargetKind = "local"
	TargetSMB    TargetKind = "smb"
	TargetS3     TargetKind = "s3"     // AWS/R2/B2/阿里/腾讯/MinIO etc. via minio-go
	TargetWebDAV TargetKind = "webdav" // Nextcloud/群晖/坚果云/Box etc.
	TargetSFTP   TargetKind = "sftp"   // any SSH-accessible host
	TargetRclone TargetKind = "rclone" // passthrough to external rclone CLI for 70+ backends
)

// ExportStatus is the lifecycle state of a single Export row.
type ExportStatus string

const (
	ExportPending ExportStatus = "pending"
	ExportRunning ExportStatus = "running"
	ExportReady   ExportStatus = "ready"
	ExportFailed  ExportStatus = "failed"
	ExportExpired ExportStatus = "expired"
)

// IntegrationExportMode controls how integrations are serialised in
// an export bundle. "plaintext" is opt-in and only meaningful for
// system integrations whose plaintext key is cached locally; for
// the rest the manifest records the field as unrecoverable.
type IntegrationExportMode string

const (
	IntegrationExportNone      IntegrationExportMode = "none"
	IntegrationExportMetadata  IntegrationExportMode = "metadata"
	IntegrationExportPlaintext IntegrationExportMode = "plaintext"
)

// Backup is the public view of one backup row.
type Backup struct {
	ID         string  `json:"id"`
	ScheduleID *string `json:"schedule_id,omitempty"`
	TargetID   string  `json:"target_id"`
	// GroupID correlates the rows produced by one fan-out invocation —
	// the same bundle written to multiple targets. Empty for a plain
	// single-target backup.
	GroupID         string       `json:"group_id,omitempty"`
	Status          BackupStatus `json:"status"`
	TriggeredBy     TriggeredBy  `json:"triggered_by"`
	Kind            BackupKind   `json:"kind"`
	StartedAt       time.Time    `json:"started_at"`
	FinishedAt      *time.Time   `json:"finished_at,omitempty"`
	Bytes           int64        `json:"bytes"`
	SHA256          string       `json:"sha256,omitempty"`
	Encrypted       bool         `json:"encrypted"`
	KeyFingerprint  string       `json:"key_fingerprint,omitempty"`
	TargetPath      string       `json:"target_path,omitempty"`
	PGVersion       string       `json:"pg_version,omitempty"`
	OpendrayVersion string       `json:"opendray_version,omitempty"`
	GitSHA          string       `json:"git_sha,omitempty"`
	Error           string       `json:"error,omitempty"`
	VerifiedAt      *time.Time   `json:"verified_at,omitempty"`
	VerifyError     string       `json:"verify_error,omitempty"`
	// ContentHash is the sha256 of the plaintext bundle; identical data
	// across runs yields the same hash, which content-dedup keys on.
	ContentHash string `json:"content_hash,omitempty"`
	// Deduped is true when this row reused a prior identical blob on the
	// same target instead of uploading a fresh copy.
	Deduped  bool           `json:"deduped,omitempty"`
	Metadata map[string]any `json:"metadata,omitempty"`
}

// Schedule is a recurring backup spec.
type Schedule struct {
	ID       string `json:"id"`
	TargetID string `json:"target_id"`
	// TargetIDs is the full set of destinations this schedule fans out
	// to (3-2-1). Always contains TargetID as its first element; a
	// single-target schedule has exactly one entry.
	TargetIDs   []string   `json:"target_ids"`
	Kind        BackupKind `json:"kind"`
	IntervalSec int        `json:"interval_sec"`
	Retention   int        `json:"retention"`
	Enabled     bool       `json:"enabled"`
	LastRunAt   *time.Time `json:"last_run_at,omitempty"`
	NextRunAt   time.Time  `json:"next_run_at"`
	CreatedAt   time.Time  `json:"created_at"`
	UpdatedAt   time.Time  `json:"updated_at"`
}

// BackupHealth is an at-a-glance roll-up the dashboard renders as a
// health strip: when the last good backup landed, plus counts of
// things that currently need an operator's attention. All fields are
// derived (no stored row) and cheap to recompute on every poll.
type BackupHealth struct {
	LastSuccessAt    *time.Time `json:"last_success_at,omitempty"`
	LastSuccessID    string     `json:"last_success_id,omitempty"`
	RecentFailures   int        `json:"recent_failures"`   // failed runs in the last 24h
	VerifyFailures   int        `json:"verify_failures"`   // succeeded backups whose last restore-verify failed
	OverdueSchedules int        `json:"overdue_schedules"` // enabled schedules >5min past their next_run_at
	Schedules        int        `json:"schedules"`         // total schedules
	EnabledSchedules int        `json:"enabled_schedules"` // enabled schedules
}

// TargetSpec is the public view of a stored BackupTarget config.
// Sensitive fields inside Config (e.g. SMB password) are returned
// redacted; the raw form is only used internally.
type TargetSpec struct {
	ID        string         `json:"id"`
	Kind      TargetKind     `json:"kind"`
	Config    map[string]any `json:"config"`
	Enabled   bool           `json:"enabled"`
	CreatedAt time.Time      `json:"created_at"`
	UpdatedAt time.Time      `json:"updated_at"`
}

// Export is the public view of one export row.
type Export struct {
	ID            string       `json:"id"`
	Status        ExportStatus `json:"status"`
	RequestedBy   string       `json:"requested_by"`
	Scope         ExportScope  `json:"scope"`
	StartedAt     time.Time    `json:"started_at"`
	FinishedAt    *time.Time   `json:"finished_at,omitempty"`
	ExpiresAt     time.Time    `json:"expires_at"`
	Bytes         int64        `json:"bytes"`
	SHA256        string       `json:"sha256,omitempty"`
	DownloadToken string       `json:"download_token,omitempty"` // omitted in list responses
	Error         string       `json:"error,omitempty"`
}

// ExportScope captures what the operator asked to be included in a
// bundle. Stored verbatim in exports.scope (JSONB).
type ExportScope struct {
	Memories     bool                  `json:"memories"`
	Integrations IntegrationExportMode `json:"integrations"`
	CustomTasks  bool                  `json:"custom_tasks"`
}

// ImportStatus is the lifecycle state of a single Import row.
type ImportStatus string

const (
	ImportPending   ImportStatus = "pending"
	ImportRunning   ImportStatus = "running"
	ImportSucceeded ImportStatus = "succeeded"
	ImportFailed    ImportStatus = "failed"
)

// EntityCounts tracks per-entity import outcomes. Each kind reports
// how many rows were created (newly inserted), skipped (already
// existed by id or unique key), and failed (e.g. constraint violation
// not covered by skip path).
type EntityCounts struct {
	Created int `json:"created"`
	Skipped int `json:"skipped"`
	Failed  int `json:"failed"`
}

// ImportCounts is the aggregate result of one ImportBundle run.
type ImportCounts struct {
	Memories     EntityCounts `json:"memories"`
	Integrations EntityCounts `json:"integrations"`
	CustomTasks  EntityCounts `json:"custom_tasks"`
}

// Import is the public view of one import row.
type Import struct {
	ID             string       `json:"id"`
	Status         ImportStatus `json:"status"`
	RequestedBy    string       `json:"requested_by"`
	StartedAt      time.Time    `json:"started_at"`
	FinishedAt     *time.Time   `json:"finished_at,omitempty"`
	SourceFilename string       `json:"source_filename,omitempty"`
	SourceBytes    int64        `json:"source_bytes"`
	Counts         ImportCounts `json:"counts"`
	Error          string       `json:"error,omitempty"`
}

// RestoreResult is what the /backups/restore endpoint returns. It
// is NOT persisted — restore is a one-shot operator operation whose
// outcome is the database itself. Audit-logged via slog.
type RestoreResult struct {
	Manifest        BundleManifest `json:"manifest"`
	BytesRead       int64          `json:"bytes_read"`
	TargetDSNUsed   string         `json:"target_dsn_used"`   // redacted (host/db only)
	FingerprintOK   bool           `json:"fingerprint_ok"`    // matched server cipher
	PGRestoreOutput string         `json:"pg_restore_output"` // tail of stderr/stdout
	Plan            RestorePlan    `json:"plan"`
	StartedAt       time.Time      `json:"started_at"`
	FinishedAt      time.Time      `json:"finished_at"`
}

// RestorePlan describes what a restore would do (dry-run) or did
// (apply). A dry-run never writes a file, never runs pg_restore and
// never takes a safety snapshot — it only reports what is in the
// bundle and where each component would land.
type RestorePlan struct {
	DryRun           bool     `json:"dry_run"`
	DumpPresent      bool     `json:"dump_present"`
	DumpBytes        int64    `json:"dump_bytes"`
	ConfigPath       string   `json:"config_path,omitempty"`  // where config.toml would land ("" = nowhere to put it)
	SecretsPath      string   `json:"secrets_path,omitempty"` // where secrets.env would land
	VaultRoots       []string `json:"vault_roots,omitempty"`  // logical roots present in vault.tar
	VaultFiles       int      `json:"vault_files"`            // file count in vault.tar
	SafetySnapshotID string   `json:"safety_snapshot_id,omitempty"`
	Applied          []string `json:"applied,omitempty"` // components actually written (apply mode)
}

// Sentinel errors. All errors returned across package boundaries
// wrap one of these so callers can errors.Is them.
var (
	ErrCipherUnconfigured = errors.New("backup: OPENDRAY_BACKUP_KEY not set")
	ErrCipherCorrupted    = errors.New("backup: ciphertext corrupted or truncated")
	ErrCipherWrongKey     = errors.New("backup: wrong key or tampered ciphertext")
	ErrCipherFormat       = errors.New("backup: unrecognised cipher format/version")

	ErrTargetNotFound       = errors.New("backup: target not found")
	ErrTargetUnsupported    = errors.New("backup: target kind not supported in this build")
	ErrBackupNotFound       = errors.New("backup: backup not found")
	ErrScheduleNotFound     = errors.New("backup: schedule not found")
	ErrExportNotFound       = errors.New("backup: export not found")
	ErrExportExpired        = errors.New("backup: export expired")
	ErrInvalidDownloadToken = errors.New("backup: invalid download token")

	ErrPgDumpUnavailable    = errors.New("backup: pg_dump binary not found on PATH")
	ErrPgRestoreUnavailable = errors.New("backup: pg_restore binary not found on PATH")
	ErrFeatureDisabled      = errors.New("backup: feature disabled (cfg.backup.enabled=false)")

	ErrImportNotFound             = errors.New("backup: import not found")
	ErrRestoreFingerprintMismatch = errors.New("backup: restore: bundle fingerprint does not match running cipher")
	ErrRestoreNoDump              = errors.New("backup: restore: no dump.bin in bundle")
	ErrImportBadBundle            = errors.New("backup: import: malformed bundle")
)

File: app/shared/src/lib/backup.ts
// Client for the /api/v1/backups, /backup-schedules, /backup-targets,
// /backup-status endpoints. Mirrors the Go shapes in internal/backup.
//
// Backup endpoints are admin-only; this module assumes the caller is
// authenticated. 404 on /backup-status means the feature is disabled
// at the server level (OPENDRAY_BACKUP_ENABLED + OPENDRAY_BACKUP_KEY
// are both required to turn it on).

import { api } from './api'

export type BackupStatus =
  | 'pending'
  | 'running'
  | 'succeeded'
  | 'failed'
  | 'deleted'

export type TriggeredBy =
  | 'scheduler'
  | 'manual'
  | 'api'
  | 'pre_migrate'
  | 'pre_restore'

// BackupKind is how much of an instance a backup captures: a plain
// encrypted pg_dump ('db_only', the default), or a 'full_instance'
// bundle that also carries the vault (notes/skills/mcp), secrets.env
// and config.toml — everything needed to rebuild a working instance.
export type BackupKind = 'db_only' | 'full_instance'

export type TargetKind =
  | 'local'
  | 'smb'
  | 's3'
  | 'webdav'
  | 'sftp'
  | 'rclone'

// TARGET_KINDS is the canonical UI ordering shown in the picker.
// Local first (default), then off-host options in increasing
// "external dependency" weight.
export const TARGET_KINDS: {
  kind: TargetKind
  label: string
  description: string
  examples: string
}[] = [
  {
    kind: 'local',
    label: 'Local disk',
    description: 'Folder on the machine running opendray',
    examples: '~/.opendray/backups · /var/backups · external HDD mounted to a path',
  },
  {
    kind: 'smb',
    label: 'SMB / CIFS',
    description: 'Windows shares + most home NAS appliances',
    examples: 'Synology · QNAP · UNAS · Windows file shares',
  },
  {
    kind: 's3',
    label: 'S3-compatible',
    description: 'AWS S3 + every S3-compatible service',
    examples: 'AWS S3 · Cloudflare R2 · Backblaze B2 · MinIO · Alibaba Cloud OSS (阿里云 OSS) · Tencent Cloud COS (腾讯云 COS) · DigitalOcean Spaces · Wasabi',
  },
  {
    kind: 'webdav',
    label: 'WebDAV',
    description: 'Self-hosted clouds + file-sharing services',
    examples: 'Nextcloud · ownCloud · Synology DSM (群晖 DSM) · Box.com · Jianguoyun (坚果云)',
  },
  {
    kind: 'sftp',
    label: 'SFTP / SSH',
    description: 'Any SSH-accessible server (Linux VPS, etc.)',
    examples: 'Hetzner Storage Box · self-hosted VPS · home Linux server',
  },
  {
    kind: 'rclone',
    label: 'rclone passthrough',
    description: 'Tap into 70+ backends configured via the rclone CLI',
    examples: 'Google Drive · OneDrive · Dropbox · Mega · pCloud · iCloud-WebDAV · Baidu Pan (百度网盘) · Aliyun Drive (阿里云盘)',
  },
]

export interface Backup {
  id: string
  schedule_id?: string | null
  target_id: string
  /** Correlates rows from one fan-out invocation (same bundle, many
   * targets). Empty for a plain single-target backup. */
  group_id?: string
  status: BackupStatus
  triggered_by: TriggeredBy
  kind: BackupKind
  started_at: string
  finished_at?: string | null
  bytes: number
  sha256?: string
  encrypted: boolean
  key_fingerprint?: string
  target_path?: string
  pg_version?: string
  opendray_version?: string
  git_sha?: string
  error?: string
  verified_at?: string | null
  verify_error?: string
  /** True when this backup reused a prior identical blob (content-dedup)
   * instead of uploading a fresh copy. */
  deduped?: boolean
  content_hash?: string
  metadata?: Record<string, unknown>
}

export interface Schedule {
  id: string
  target_id: string
  /** Full fan-out destination set (3-2-1). Always includes target_id
   * as its first element; a single-target schedule has one entry. */
  target_ids: string[]
  kind: BackupKind
  interval_sec: number
  retention: number
  enabled: boolean
  last_run_at?: string | null
  next_run_at: string
  created_at: string
  updated_at: string
}

export interface TargetSpec {
  id: string
  kind: TargetKind
  config: Record<string, unknown>
  enabled: boolean
  created_at: string
  updated_at: string
}

// Since PR #49 /backup-status always returns 200 — `enabled` tells
// you whether the feature is actively running, `configured` whether
// a passphrase is available from any source, and `requires_restart`
// whether the operator just wrote a key file via the UI and the
// gateway hasn't picked it up yet. The legacy `ok` field is still
// here for the live-feature pg_dump health check.
export interface BackupStatusReport {
  enabled: boolean
  configured: boolean
  configured_via: 'env' | 'file' | ''
  can_disable_via_ui: boolean
  requires_restart: boolean
  key_file_path: string
  // Populated only when enabled === true.
  ok?: boolean
  key_fingerprint?: string
  pg_dump_version?: string
  pg_dump_error?: string
  pg_restore_version?: string
}

export async function fetchBackupStatus(): Promise<BackupStatusReport> {
  return api<BackupStatusReport>('/api/v1/backup-status')
}

// BackupHealth is the at-a-glance roll-up the dashboard renders as a
// health strip. Mirrors backup.BackupHealth in Go. All counts are
// "needs attention" signals — non-zero means something to look at.
export interface BackupHealth {
  last_success_at?: string | null
  last_success_id?: string
  recent_failures: number // failed runs in the last 24h
  verify_failures: number // succeeded backups whose last restore-verify failed
  overdue_schedules: number // enabled schedules >5min past their next_run_at
  schedules: number // total schedules
  enabled_schedules: number // enabled schedules
}

export async function fetchBackupHealth(): Promise<BackupHealth> {
  return api<BackupHealth>('/api/v1/backup-health')
}

export interface BackupSetupResult {
  ok: boolean
  key_file_path: string
  requires_restart: boolean
  /** Server-generated passphrase; only present when mode==='generate'. */
  passphrase?: string
}

/** POST /backup-setup. Writes the key file; returns the passphrase
 * once when mode==='generate'. The operator MUST save it before
 * continuing — no recovery path exists. */
export async function postBackupSetup(opts: {
  mode: 'generate' | 'paste'
  passphrase?: string
}): Promise<BackupSetupResult> {
  return api<BackupSetupResult>('/api/v1/backup-setup', {
    method: 'POST',
    body: JSON.stringify(opts),
    headers: { 'Content-Type': 'application/json' },
  })
}

/** POST /backup-setup/disable. Removes the key file (no-op for env-
 * configured deployments — server rejects with 409 in that case). */
export async function postBackupDisable(): Promise<{
  ok: boolean
  requires_restart: boolean
}> {
  return api<{ ok: boolean; requires_restart: boolean }>(
    '/api/v1/backup-setup/disable',
    { method: 'POST' },
  )
}

export async function listBackups(opts?: {
  status?: BackupStatus
  targetId?: string
  limit?: number
}): Promise<Backup[]> {
  const params = new URLSearchParams()
  if (opts?.status) params.set('status', opts.status)
  if (opts?.targetId) params.set('target_id', opts.targetId)
  if (opts?.limit) params.set('limit', String(opts.limit))
  const q = params.toString()
  const res = await api<{ backups: Backup[] }>(
    `/api/v1/backups${q ? `?${q}` : ''}`,
  )
  return res.backups
}

export async function getBackup(id: string): Promise<Backup> {
  return api<Backup>(`/api/v1/backups/${encodeURIComponent(id)}`)
}

export async function createBackup(opts: {
  targetId?: string
  /** Fan-out destinations (3-2-1). When set, takes precedence over
   * targetId and the same bundle is written to every target. */
  targetIds?: string[]
  kind?: BackupKind
  includeConfig?: boolean
}): Promise<Backup> {
  return api<Backup>('/api/v1/backups', {
    method: 'POST',
    body: {
      ...(opts.targetIds && opts.targetIds.length > 0
        ? { target_ids: opts.targetIds }
        : { target_id: opts.targetId ?? 'local' }),
      kind: opts.kind ?? 'db_only',
      include_config: opts.includeConfig ?? false,
    },
  })
}

export async function deleteBackup(id: string): Promise<void> {
  await api<unknown>(`/api/v1/backups/${encodeURIComponent(id)}`, {
    method: 'DELETE',
  })
}

/** Browser-friendly download URL. Auth rides in the query string — a
 *  browser anchor navigation can't set an Authorization header, and there
 *  is no auth cookie, so the admin bearer must go via the `?token=`
 *  fallback the gateway middleware accepts (same path fs/export downloads
 *  use). Without it the endpoint returns {"error":"unauthorized"}. */
export function backupDownloadURL(id: string, token: string): string {
  return `/api/v1/backups/${encodeURIComponent(id)}/download?token=${encodeURIComponent(token)}`
}

export async function listTargets(): Promise<TargetSpec[]> {
  const res = await api<{ targets: TargetSpec[] }>('/api/v1/backup-targets')
  return res.targets
}

export async function createTarget(opts: {
  id?: string
  kind: TargetKind
  config: Record<string, unknown>
  enabled: boolean
}): Promise<TargetSpec> {
  return api<TargetSpec>('/api/v1/backup-targets', {
    method: 'POST',
    body: opts,
  })
}

export async function updateTarget(
  id: string,
  patch: { config?: Record<string, unknown>; enabled?: boolean },
): Promise<TargetSpec> {
  return api<TargetSpec>(`/api/v1/backup-targets/${encodeURIComponent(id)}`, {
    method: 'PATCH',
    body: patch,
  })
}

export async function deleteTarget(id: string): Promise<void> {
  await api<unknown>(`/api/v1/backup-targets/${encodeURIComponent(id)}`, {
    method: 'DELETE',
  })
}

export interface TestTargetResult {
  ok: boolean
  error?: string
}

export async function testTarget(id: string): Promise<TestTargetResult> {
  return api<TestTargetResult>(
    `/api/v1/backup-targets/${encodeURIComponent(id)}/test`,
    { method: 'POST' },
  )
}

// ── schedules ────────────────────────────────────────────────────

export async function listSchedules(): Promise<Schedule[]> {
  const res = await api<{ schedules: Schedule[] }>('/api/v1/backup-schedules')
  return res.schedules
}

export async function createSchedule(opts: {
  targetId?: string
  /** Fan-out destination set (3-2-1). When set, its first element
   * becomes the primary target_id. */
  targetIds?: string[]
  kind?: BackupKind
  intervalSec: number
  retention: number
  enabled: boolean
}): Promise<Schedule> {
  return api<Schedule>('/api/v1/backup-schedules', {
    method: 'POST',
    body: {
      ...(opts.targetIds && opts.targetIds.length > 0
        ? { target_ids: opts.targetIds }
        : { target_id: opts.targetId }),
      kind: opts.kind ?? 'db_only',
      interval_sec: opts.intervalSec,
      retention: opts.retention,
      enabled: opts.enabled,
    },
  })
}

export async function updateSchedule(
  id: string,
  patch: {
    kind?: BackupKind
    targetIds?: string[]
    intervalSec?: number
    retention?: number
    enabled?: boolean
  },
): Promise<Schedule> {
  return api<Schedule>(`/api/v1/backup-schedules/${encodeURIComponent(id)}`, {
    method: 'PATCH',
    body: {
      ...(patch.kind !== undefined && { kind: patch.kind }),
      ...(patch.targetIds !== undefined && { target_ids: patch.targetIds }),
      ...(patch.intervalSec !== undefined && {
        interval_sec: patch.intervalSec,
      }),
      ...(patch.retention !== undefined && { retention: patch.retention }),
      ...(patch.enabled !== undefined && { enabled: patch.enabled }),
    },
  })
}

export async function deleteSchedule(id: string): Promise<void> {
  await api<unknown>(`/api/v1/backup-schedules/${encodeURIComponent(id)}`, {
    method: 'DELETE',
  })
}

// ── helpers ──────────────────────────────────────────────────────

// ── exports (Plan C) ─────────────────────────────────────────────

export type IntegrationExportMode = 'none' | 'metadata' | 'plaintext'

export type ExportStatus =
  | 'pending'
  | 'running'
  | 'ready'
  | 'failed'
  | 'expired'

export interface ExportScope {
  memories: boolean
  integrations: IntegrationExportMode
  custom_tasks: boolean
}

export interface ExportRecord {
  id: string
  status: ExportStatus
  requested_by: string
  scope: ExportScope
  started_at: string
  finished_at?: string | null
  expires_at: string
  bytes: number
  sha256?: string
  download_token?: string
  error?: string
}

export async function listExports(): Promise<ExportRecord[]> {
  const res = await api<{ exports: ExportRecord[] }>('/api/v1/exports')
  return res.exports
}

export async function getExport(id: string): Promise<ExportRecord> {
  return api<ExportRecord>(`/api/v1/exports/${encodeURIComponent(id)}`)
}

export async function createExport(opts: {
  memories: boolean
  integrations: IntegrationExportMode
  customTasks: boolean
}): Promise<ExportRecord> {
  return api<ExportRecord>('/api/v1/exports', {
    method: 'POST',
    body: {
      memories: opts.memories,
      integrations: opts.integrations,
      custom_tasks: opts.customTasks,
    },
  })
}

export async function deleteExport(id: string): Promise<void> {
  await api<unknown>(`/api/v1/exports/${encodeURIComponent(id)}`, {
    method: 'DELETE',
  })
}

export function exportDownloadURL(id: string, token: string): string {
  return `/api/v1/exports/${encodeURIComponent(id)}/download?token=${encodeURIComponent(token)}`
}

// ── inventory (what's in a backup) ──────────────────────────────

export interface InventoryTable {
  name: string
  count: number
}

export interface InventoryGroup {
  id: string
  label: string
  description: string
  tables: InventoryTable[]
}

export async function fetchBackupInventory(): Promise<InventoryGroup[]> {
  const res = await api<{ groups: InventoryGroup[] }>(
    '/api/v1/backup-inventory',
  )
  return res.groups
}

// ── restore (A) ──────────────────────────────────────────────────

// RestorePlan describes what a restore would do (dry-run) or did
// (apply). Mirrors backup.RestorePlan in Go.
export interface RestorePlan {
  dry_run: boolean
  dump_present: boolean
  dump_bytes: number
  config_path?: string
  secrets_path?: string
  vault_roots?: string[]
  vault_files: number
  safety_snapshot_id?: string
  applied?: string[]
}

export interface RestoreResult {
  manifest: {
    version: string
    backup_id: string
    created_at: string
    opendray_version?: string
    pg_version?: string
    encryption: { algo: string; fingerprint: string }
  }
  bytes_read: number
  target_dsn_used: string
  fingerprint_ok: boolean
  pg_restore_output: string
  plan: RestorePlan
  started_at: string
  finished_at: string
}

// restoreBackup defaults to a DRY RUN: it validates the bundle and
// returns a plan without changing anything. Pass apply=true (plus the
// confirm phrase) to commit; force=true proceeds even if the
// pre-restore safety snapshot fails.
export async function restoreBackup(opts: {
  bundle: File
  targetDsn?: string
  clean: boolean
  apply?: boolean
  force?: boolean
  confirm?: string
  note?: string
}): Promise<RestoreResult> {
  const fd = new FormData()
  fd.set('bundle', opts.bundle)
  if (opts.targetDsn) fd.set('target_dsn', opts.targetDsn)
  fd.set('clean', String(opts.clean))
  fd.set('apply', String(opts.apply ?? false))
  if (opts.force) fd.set('force', 'true')
  if (opts.confirm) fd.set('confirm', opts.confirm)
  if (opts.note) fd.set('note', opts.note)
  return api<RestoreResult>('/api/v1/backups/restore', {
    method: 'POST',
    body: fd,
  })
}

// ── recovery kit ─────────────────────────────────────────────────

export interface RecoveryKit {
  version: number
  created_at: string
  key_fingerprint: string
  wrapped_key: string
}

/** POST /backup-recovery-kit. Wraps the backup passphrase under a
 *  separate recovery passphrase the operator stores out-of-band. The
 *  returned kit + that recovery passphrase are what recover a dead
 *  host (see `opendray recover-key`). */
export async function fetchRecoveryKit(
  recoveryPassphrase: string,
): Promise<RecoveryKit> {
  return api<RecoveryKit>('/api/v1/backup-recovery-kit', {
    method: 'POST',
    body: { recovery_passphrase: recoveryPassphrase },
  })
}

// ── imports (C reverse) ──────────────────────────────────────────

export type ImportStatus = 'pending' | 'running' | 'succeeded' | 'failed'

export interface EntityCounts {
  created: number
  skipped: number
  failed: number
}

export interface ImportRecord {
  id: string
  status: ImportStatus
  requested_by: string
  started_at: string
  finished_at?: string | null
  source_filename?: string
  source_bytes: number
  counts: {
    memories: EntityCounts
    integrations: EntityCounts
    custom_tasks: EntityCounts
  }
  error?: string
}

export async function listImports(limit = 20): Promise<ImportRecord[]> {
  const res = await api<{ imports: ImportRecord[] }>(
    `/api/v1/imports?limit=${limit}`,
  )
  return res.imports
}

export async function getImport(id: string): Promise<ImportRecord> {
  return api<ImportRecord>(`/api/v1/imports/${encodeURIComponent(id)}`)
}

export async function createImport(opts: {
  bundle: File
  memories: boolean
  integrations: boolean
  customTasks: boolean
}): Promise<ImportRecord> {
  const fd = new FormData()
  fd.set('bundle', opts.bundle)
  fd.set('memories', String(opts.memories))
  fd.set('integrations', String(opts.integrations))
  fd.set('custom_tasks', String(opts.customTasks))
  return api<ImportRecord>('/api/v1/imports', {
    method: 'POST',
    body: fd,
  })
}

// ── helpers ──────────────────────────────────────────────────────

export function formatBytes(n: number): string {
  if (n < 1024) return `${n} B`
  if (n < 1024 * 1024) return `${(n / 1024).toFixed(1)} KiB`
  if (n < 1024 * 1024 * 1024) return `${(n / 1024 / 1024).toFixed(1)} MiB`
  return `${(n / 1024 / 1024 / 1024).toFixed(2)} GiB`
}

export function formatInterval(sec: number): string {
  if (sec < 60) return `${sec}s`
  if (sec < 3600) return `${Math.round(sec / 60)} min`
  if (sec < 86400) return `${Math.round(sec / 3600)} h`
  return `${Math.round(sec / 86400)} d`
}
